src/main/java/io/sqreen/powerwaf/CharSequenceWrapper.java

```diff
@@ -0,0 +1,57 @@
+package io.sqreen.powerwaf;
+
+import java.lang.reflect.Field;
+import java.nio.Buffer;
+import java.nio.CharBuffer;
+
+public class CharSequenceWrapper {
+    private static final Field str;
+    private static final Field offset;
+    private static final Field mark;
+    private static final Field position;
+    private static final Field limit;
+    private static final Field capacity;
+    private static boolean resetBuffer;
+
+    static {
+        str = getField(CharBuffer.wrap("").getClass(), "str");
+        offset = getField(CharBuffer.class, "offset");
+        mark = getField(Buffer.class, "mark");
+        position = getField(Buffer.class, "position");
+        limit = getField(Buffer.class, "limit");
+        capacity = getField(Buffer.class, "capacity");
+        resetBuffer = str != null && offset != null && mark != null && position != null
+                && limit != null && capacity != null;
+    }
+
+    private static Field getField(Class<?> klass, String name) {
+        try {
+            Field f = klass.getDeclaredField(name);
+            f.setAccessible(true);
+            return f;
+        } catch (Throwable t) {
+            return null;
+        }
+    }
+
+    public static CharBuffer wrap(CharSequence cs, CharBuffer previous) {
+        if (!resetBuffer || previous == null || cs == null) {
+            return CharBuffer.wrap(cs);
+        } else {
+            // we're assuming that this is a StringCharBuffer until proven otherwise
```

Comment:
This is for any case where the `CharBuffer` is not a `StringCharBuffer` in which case we can't really say much about it. We can't do a instance of since the class is only package reachable.